File: src/audit/kernel/verify.ts
/**
 * Standalone chain verifier — recomputes the hash-chain, detects tamper/reorder/gap, and verifies
 * the Ed25519 checkpoint signature. It trusts ONLY the entries + checkpoint it is handed (plus the
 * public key); it does not read a log instance's internal state. It shares the pinned hash/frame
 * functions with the log on purpose — recomputing with the SAME definition is the point.
 */
import { type KeyObject, verify } from "node:crypto";
import { GENESIS_PREV_HASH, type SignedChain, checkpointBytes, computeEntryHash } from "./log.js";

export type VerifyResult =
  | { ok: true; length: number }
  | { ok: false; brokenAt: number; reason: string };

export function verifyChain(chain: SignedChain, publicKey: KeyObject): VerifyResult {
  const { entries, checkpoint } = chain;

  let prevHash = GENESIS_PREV_HASH;
  for (let i = 0; i < entries.length; i++) {
    const entry = entries[i];
    if (entry === undefined) {
      return { ok: false, brokenAt: i, reason: "missing entry" };
    }
    if (entry.sequence !== i) {
      return {
        ok: false,
        brokenAt: i,
        reason: `sequence not monotonic: expected ${i}, got ${entry.sequence}`,
      };
    }
    if (entry.prevHash !== prevHash) {
      return { ok: false, brokenAt: i, reason: "prev-hash linkage broken (reorder/insert/tamper)" };
    }
    if (computeEntryHash(entry.event, entry.prevHash, entry.sequence) !== entry.entryHash) {
      return { ok: false, brokenAt: i, reason: "entry hash mismatch (tampered content)" };
    }
    prevHash = entry.entryHash;
  }

  if (checkpoint.length !== entries.length) {
    return { ok: false, brokenAt: entries.length, reason: "checkpoint length mismatch" };
  }
  const head = entries.at(-1);
  const headEntryHash = head ? head.entryHash : GENESIS_PREV_HASH;
  if (checkpoint.headEntryHash !== headEntryHash) {
    return { ok: false, brokenAt: entries.length, reason: "checkpoint head mismatch" };
  }
  const signatureOk = verify(
    null,
    checkpointBytes(checkpoint.headEntryHash, checkpoint.length),
    publicKey,
    Buffer.from(checkpoint.signature, "base64"),
  );
  if (!signatureOk) {
    return { ok: false, brokenAt: entries.length, reason: "checkpoint signature invalid" };
  }

  return { ok: true, length: entries.length };
}

File: kernel/internal/verify/verify.go
// Package verify is a standalone chain verifier: it recomputes the hash-chain, detects
// tamper/reorder/gap, and verifies the Ed25519 checkpoint. It trusts ONLY the entries + checkpoint
// + public key it is handed — it does NOT import internal/log (depguard enforces this), so "the
// verifier is independent of the producer" is true by construction. It shares the pinned
// hash/frame functions (internal/chain) with the log on purpose: recomputing with the SAME
// definition is the point. Mirrors src/audit/kernel/verify.ts byte-for-byte in Ok/BrokenAt.
package verify

import (
	"crypto/ed25519"
	"fmt"

	"github.com/agent-os/kernel/internal/chain"
)

// VerifyResult reports either success (Ok, Length) or the first break (BrokenAt, Reason).
// Ok and BrokenAt are the byte-for-byte contract; Reason is human-readable, not a contract surface.
type VerifyResult struct {
	Ok       bool
	Length   int
	BrokenAt int
	Reason   string
}

// VerifyChain recomputes the chain in the pinned check order: sequence -> linkage -> entryHash ->
// checkpoint.length -> head -> signature. Fails closed.
func VerifyChain(c chain.SignedChain, publicKey ed25519.PublicKey) VerifyResult {
	prevHash := chain.GenesisPrevHash
	for i := range c.Entries {
		entry := c.Entries[i]
		if entry.Sequence != i {
			return VerifyResult{BrokenAt: i, Reason: fmt.Sprintf("sequence not monotonic: expected %d, got %d", i, entry.Sequence)}
		}
		if entry.PrevHash != prevHash {
			return VerifyResult{BrokenAt: i, Reason: "prev-hash linkage broken (reorder/insert/tamper)"}
		}
		got, err := chain.ComputeEntryHash(entry.Event, entry.PrevHash, entry.Sequence)
		if err != nil || got != entry.EntryHash {
			return VerifyResult{BrokenAt: i, Reason: "entry hash mismatch (tampered content)"}
		}
		prevHash = entry.EntryHash
	}

	if c.Checkpoint.Length != len(c.Entries) {
		return VerifyResult{BrokenAt: len(c.Entries), Reason: "checkpoint length mismatch"}
	}
	headEntryHash := chain.GenesisPrevHash
	if n := len(c.Entries); n > 0 {
		headEntryHash = c.Entries[n-1].EntryHash
	}
	if c.Checkpoint.HeadEntryHash != headEntryHash {
		return VerifyResult{BrokenAt: len(c.Entries), Reason: "checkpoint head mismatch"}
	}
	if !chain.VerifyCheckpoint(publicKey, c.Checkpoint.HeadEntryHash, c.Checkpoint.Length, c.Checkpoint.Signature) {
		return VerifyResult{BrokenAt: len(c.Entries), Reason: "checkpoint signature invalid"}
	}
	return VerifyResult{Ok: true, Length: len(c.Entries)}
}
